iot-hub/Samples/module/ModuleSample/ExceptionExtensions.cs

```diff
@@ -0,0 +1,39 @@
+﻿// Microsoft.All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Azure.Devices.Client.Samples
+{
+    internal static class ExceptionExtensions
+    {
+        internal static IEnumerable<Exception> Unwind(this Exception exception, bool unwindAggregate = false)
+        {
+            while (exception != null)
+            {
+                yield return exception;
+
+                if (!unwindAggregate)
+                {
+                    exception = exception.InnerException;
```

Comment:
does this change which exception the original object object points to?

Reply:
no, it simply returns the inner exceptions for the exception, so that we can use them for verification/ mapping etc
eg case - where the outer exception is IotHubCommunicationException, but the inner exception is a TimeoutException -> the client should retry during operation timeouts.